sway/tree/container.c

```diff
@@ -1609,3 +1619,68 @@ bool container_is_scratchpad_hidden_or_child(struct sway_container *con) {
 	con = container_toplevel_ancestor(con);
 	return con->scratchpad && !con->workspace;
 }
+
+static enum sway_container_layout layout_category(enum sway_container_layout layout) {
+	switch (layout) {
+		case L_HORIZ: /* fallthrough */
+		case L_TABBED:
+			return L_HORIZ;
+		case L_VERT: /* fallthrough */
+		case L_STACKED:
+			return L_VERT;
+		case L_NONE:
+			return L_NONE;
+	}
+	return L_NONE;
+}
+
+static bool container_is_squashable(struct sway_container *con,
+		struct sway_container *child) {
+	enum sway_container_layout gp_orient = layout_category(
+			con->parent ? con->parent->layout : con->workspace->layout);
+	enum sway_container_layout con_orient = layout_category(con->layout);
+	enum sway_container_layout child_orient = layout_category(child->layout);
+
+	return (con->layout == L_HORIZ || con->layout == L_VERT) &&
+		(child->layout == L_HORIZ || child->layout == L_VERT) &&
+		con_orient != child_orient && gp_orient == child_orient;
+}
+
+static void container_squash_children(struct sway_container *con) {
+	for (int i = 0; i < con->children->length; i++) {
+		struct sway_container *child = con->children->items[i];
+		i += container_squash(child);
+	}
+}
+
+// Returns the number of new containers added to the parent
+int container_squash(struct sway_container *con) {
+	int change = 0;
+	if (!con->children) {
+		return change;
+	}
+	if (con->children->length != 1) {
+		container_squash_children(con);
+		return change;
+	}
+	struct sway_container *child = con->children->items[0];
+	int idx = container_sibling_index(con);
+	if (container_is_squashable(con, child)) {
+		while (child->children->length) {
+			struct sway_container *current = child->children->items[0];
+			container_detach(current);
+			if (con->parent) {
+				container_insert_child(con->parent, current, idx);
+			} else {
+				workspace_insert_tiling(con->workspace, current, idx,
+						L_NONE);
+			}
+			change++;
+		}
+		container_reap_empty(child);
+		change--;
+		return change;
+	}
+	container_squash_children(con);
```

Comment:
The `if` above ends with a `return`, so we can put this line in an `else` block, and deduplicate the return by removing the one on L1682 and keeping the one on L1685.